File: api/src/cron/jobs/sitespeedScrape.ts
import { eq } from 'drizzle-orm';
import { getScreenshotUrl, runSitespeedReport } from '#src/service/sitespeed';
import { getConnection, schema } from '../../db.ts';

export async function scrapeSitespeedForAllShops() {
    const drizzle = getConnection();

    console.log('Starting sitespeed scrape for all shops');

    const shops = await drizzle.query.shop.findMany({
        columns: {
            id: true,
        },
        where: eq(schema.shop.sitespeedEnabled, true),
    });

    console.log(`Found ${shops.length} shops with sitespeed enabled`);

    for (const shop of shops) {
        await scrapeSingleSitespeedShop(shop.id);
    }

    console.log('Sitespeed scrape completed for all shops');
}

export async function scrapeSingleSitespeedShop(shopId: number) {
    const drizzle = getConnection();

    const shop = await drizzle.query.shop.findFirst({
        columns: {
            id: true,
            url: true,
            name: true,
            sitespeedEnabled: true,
            sitespeedUrls: true,
            connectionIssueCount: true,
        },
        where: eq(schema.shop.id, shopId),
    });

    if (!shop) {
        console.error(`Shop ${shopId} not found`);
        return;
    }

    if (
        !shop.sitespeedEnabled ||
        !shop.sitespeedUrls ||
        shop.connectionIssueCount > 0
    ) {
        if (!shop.sitespeedUrls) {
            console.warn(
                `Shop ${shop.id} (${shop.name}) has sitespeed disabled or no URLs configured`,
            );
        }

        if (shop.connectionIssueCount > 0) {
            console.warn(
                `Shop ${shop.id} (${shop.name}) has connection issues, skipping sitespeed analysis`,
            );
        }

        return;
    }

    console.log(`Running sitespeed analysis for shop ${shop.id}: ${shop.name}`);

    shop.sitespeedUrls = shop.sitespeedUrls.map((url) =>
        url.replace('http://localhost:3889', 'http://demoshop:8000'),
    );

    try {
        const result = await runSitespeedReport(shop.id, shop.sitespeedUrls);

        // Store the metrics in the database
        await drizzle
            .insert(schema.shopSitespeed)
            .values({
                shopId: shop.id,
                createdAt: new Date(),
                ttfb: result.ttfb || null,
                fullyLoaded: result.fullyLoaded || null,
                largestContentfulPaint: result.largestContentfulPaint || null,
                firstContentfulPaint: result.firstContentfulPaint || null,
                cumulativeLayoutShift: result.cumulativeLayoutShift || null,
                transferSize: result.transferSize || null,
            })
            .execute();

        await drizzle
            .update(schema.shop)
            .set({
                shopImage: getScreenshotUrl(shop.id),
            })
            .where(eq(schema.shop.id, shop.id))
            .execute();

        console.log(`Sitespeed analysis completed for shop ${shop.id}`);
    } catch (error) {
        console.error(
            `Error running sitespeed analysis for shop ${shop.id}`,
            error,
        );
        throw error;
    }
}

File: api/internal/jobs/sitespeed_scrape.go
package jobs

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strings"

	"github.com/friendsofshopware/shopmon/api/internal/config"
	"github.com/friendsofshopware/shopmon/api/internal/database/queries"
	"github.com/friendsofshopware/shopmon/api/internal/httputil"
	"github.com/jackc/pgx/v5/pgxpool"
)

type SitespeedScrapeHandler struct {
	pool    *pgxpool.Pool
	queries *queries.Queries
	cfg     *config.Config
}

func NewSitespeedScrapeHandler(pool *pgxpool.Pool, q *queries.Queries, cfg *config.Config) *SitespeedScrapeHandler {
	return &SitespeedScrapeHandler{pool: pool, queries: q, cfg: cfg}
}

func (h *SitespeedScrapeHandler) HandleScrapeAll(ctx context.Context, _ SitespeedScrapeAll) error {
	environments, err := h.queries.GetEnvironmentsWithSitespeedEnabled(ctx)
	if err != nil {
		return fmt.Errorf("get environments with sitespeed: %w", err)
	}

	for _, env := range environments {
		if err := h.scrapeEnvironment(ctx, env); err != nil {
			slog.Error("failed to scrape sitespeed", "environmentId", env.ID, "error", err)
		}
	}
	return nil
}

func (h *SitespeedScrapeHandler) HandleScrape(ctx context.Context, msg SitespeedScrape) error {
	environments, err := h.queries.GetEnvironmentsWithSitespeedEnabled(ctx)
	if err != nil {
		return err
	}

	for _, env := range environments {
		if env.ID == msg.EnvironmentID {
			return h.scrapeEnvironment(ctx, env)
		}
	}

	return fmt.Errorf("environment %d not found or sitespeed not enabled", msg.EnvironmentID)
}

func (h *SitespeedScrapeHandler) scrapeEnvironment(ctx context.Context, env queries.GetEnvironmentsWithSitespeedEnabledRow) error {
	log := slog.With("environmentId", env.ID)

	// Get URLs to test - default is the environment URL
	var urls []string
	if env.SitespeedUrls != nil {
		if err := json.Unmarshal(env.SitespeedUrls, &urls); err != nil {
			slog.Error("failed to unmarshal sitespeed urls", "environmentId", env.ID, "error", err)
			return nil
		}
	}
	for i, u := range urls {
		urls[i] = strings.Replace(u, "http://localhost:3889", "http://demoshop:8000", 1)
	}

	recordID := fmt.Sprintf("%senvironment-%d", h.cfg.SitespeedPrefix, env.ID)

	// Call sitespeed service
	log.Info("calling sitespeed service", "urls", len(urls), "recordId", recordID)
	endpoint := fmt.Sprintf("%s/api/result/%s", h.cfg.SitespeedEndpoint, recordID)

	reqBody, _ := json.Marshal(map[string]interface{}{
		"urls": urls,
		"headers": map[string]string{
			"shopmon-shop-token": env.EnvironmentToken,
		},
	})

	req, err := http.NewRequestWithContext(ctx, "POST", endpoint, bytes.NewReader(reqBody))
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+h.cfg.SitespeedAPIKey)
	req.Header.Set("Content-Type", "application/json")

	resp, err := httputil.NewHTTPClient().Do(req)
	if err != nil {
		return fmt.Errorf("call sitespeed: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 400 {
		return fmt.Errorf("sitespeed error (%d): %s", resp.StatusCode, string(body))
	}

	// Parse response and save metrics
	log.Info("saving sitespeed metrics")
	var result struct {
		TTFB                   *int32   `json:"ttfb"`
		FullyLoaded            *int32   `json:"fullyLoaded"`
		LargestContentfulPaint *int32   `json:"largestContentfulPaint"`
		FirstContentfulPaint   *int32   `json:"firstContentfulPaint"`
		CumulativeLayoutShift  *float32 `json:"cumulativeLayoutShift"`
		TransferSize           *int32   `json:"transferSize"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		slog.Error("failed to unmarshal sitespeed result", "environmentId", env.ID, "error", err)
	}

	if err := h.queries.InsertEnvironmentSitespeed(ctx, queries.InsertEnvironmentSitespeedParams{
		EnvironmentID:          &env.ID,
		DeploymentID:           env.ActiveDeploymentID,
		Ttfb:                   result.TTFB,
		FullyLoaded:            result.FullyLoaded,
		LargestContentfulPaint: result.LargestContentfulPaint,
		FirstContentfulPaint:   result.FirstContentfulPaint,
		CumulativeLayoutShift:  result.CumulativeLayoutShift,
		TransferSize:           result.TransferSize,
	}); err != nil {
		slog.Error("failed to insert environment sitespeed", "environmentId", env.ID, "error", err)
	}

	screenshotURL := fmt.Sprintf("%s/screenshot/%s", h.cfg.SitespeedEndpoint, recordID)
	if err := h.queries.UpdateEnvironmentImage(ctx, queries.UpdateEnvironmentImageParams{
		EnvironmentImage: &screenshotURL,
		ID:               env.ID,
	}); err != nil {
		slog.Error("failed to update environment image", "environmentId", env.ID, "error", err)
	}

	log.Info("sitespeed scrape completed")
	return nil
}
